README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# billmillr

**Author:** [Matthew Hoff](https://github.com/mghoff) <br/> **License:**
[MIT](https://opensource.org/licenses/MIT)<br/>

[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](http://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/ropensci/epubr?branch=master&svg=true)](https://ci.appveyor.com/project/leonawicz/epubr)

## The Premise

This package provides functions & documentation for solving “The Bill
Miller Problem” presented within the book entitled *The Drunkard’s Walk:
How Randomness Rules Our Lives*, written by Leonard Mlodinow. More
generally, the functions herein can be used to solve for, either
analytically or by simulation, the likelihood of obtaining a winning
streak of given length within a given number of attempts, attempted by a
specified number of individuals.

The premise of this story goes that Bill Miller (financier) was an
amazing stock picker after having performed incredibly well - beating
the market - defined as outperforming the S&P500 - each year over 15
consecutive years. As a result, he was celebrated and acclaimed by the
likes of Forbes and others, who claimed that the likelihood of his
ability to perform this well was 1 in 32,768, or \~0.0032%, which is
roughly true is one considers only the individual, Bill Miller, picking
stocks.

However, what Dr. Mlodinow understands is that there are many hedge
funds all picking stocks and based on this fact, poses the question:
“Out of 1000 stock pickers (coin tossers), what are the odds that 1 of
them beats the market over 15 consecutive years?” The answer to which is
roughly 3%.

Dr. Mlodinow then further refines this calculation by considering the
scenario of beating the market 15 years consecutively or longer over a
40 year period; i.e. over 40 years and with 1000 traders, what is the
probability that at least 1 trader will obtain a winning streak of at
least 15 years with the odds of winning in a given year equal to 0.5.
Based on this refinement, Dr. Mlodinow claims the odds are roughly 3 out
of 4, or 75%; however, he provides no proof or evidence of this claim.

**The resulting likelihood for any one person to beat the market within
a definite start and stop time of at least 15 years in a single 40 year
period, while larger than the 3% of his first calculation, is much
smaller than the final proposed result of \~75%, which this package has
been built to answer and results in a value of \~33%.**

## The Math

#### Part 1:

One must compute the odds of getting a run (Streak) of at least k heads
out of N coin tosses where p (q = 1-p) is the probability of obtaining
heads (tails) from the toss of a coin.

Mathematically,

![](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;S%5BN%2C%20K%5D%20%3D%20p%5Ek%20%2B%20%5Csum_%7Bj%3D1%2C%20K%7D%20%5C%7B%20p%5E%7B%28j-1%29%7D%20%281-p%29%20S%5BN-j%2C%20K%5D%20%5C%7D)

which can be broken down recursively into the sum of terms:

![](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;S%5Bn%2C%20k%5D%20%3D%20p%5Ek%20%2B%20...%20%3D%20%5Csum_%7Bj%3D1%2C%20k%7D%20%5C%7B%20p%5E%7B%28j-1%29%7D%20%281-p%29%20S%5Bn-j%2C%20k%5D%20%5C%7D%20%5Ctext%7B%20for%20%7D%201%20%5Cle%20j%20%5Cle%20k)

which is provided by `odds_of_streak()`.

For more information on the math behind this recursive odds calculation,
see this [Ask A
Mathematician](https://www.askamathematician.com/2010/07/q-whats-the-chance-of-getting-a-run-of-k-successes-in-n-bernoulli-trials-why-use-approximations-when-the-exact-answer-is-known/)
post.

#### Part 2:

To calculate the likelihood that at least k out of M people will obtain
a streak of at least j heads out of N coin tosses, one must perform the
following:

1.  Calculate the PDF:

![](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cmathrm%7BP%7D%28M%20%3D%20k%29%20%3D%20%7BM%20%5Cchoose%20k%7Dp%5E%7Bk%7D%281-p%29%5E%7B%28M-k%29%7D)

Again, this is provided by `odds_of_streak()`.

2.  Then, calculate the CDF:

![](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cmathrm%7BP%7D%28X%20%5Cle%20x%29%20%3D%20%5Csum_%7Bi%3D0%2Cx%7D%20%5Cmathrm%7Bpdf%7D%20%5Ctext%7B%20for%20%7D%20%7Bi%20%5Cle%20x%7D)

3.  And, finally, calculate the final result:

![](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Cmathrm%7BP%7D%28X%20%3E%20x%29%20%3D%201%20-%20%5Cmathrm%7BP%7D%28X%20%5Cle%20x%29%20%5Ctext%7B%3B%20i.e.%20%7D%201%20-%20%282%29)

which is provided by `prob_of_at_least_k()`.

### Example 1: Mathematical Proof

Load Package…

``` r
library(billmillr)
```

Calculate the likelihood of obtaining a winning streak of at least 3
heads out of 5 coin tosses given that the probability p (q) of heads
(tails) is fair, i.e. p = q = 0.5.

``` r
pS <- odds_of_streak(num_coins = 5, min_heads = 3, prob_heads = 0.5)
pS
#> [1] 0.25

### NOTE:
# Example from "The Drunkard's Walk: How Randomness Rules Our Lives"
# tictoc::tic()
# odds_of_streak(40, 15, 0.5) = 0.000411981
# tictoc::toc() # 24713.89 sec elapsed (6.865 hours)
```

Now calculate the probability that at least 1 person out of 8 people
will obtain such a winning streak of 3 heads given that the probability
of said streak is 0.25.

``` r
pK <- prob_of_at_least_k(N = 8, K = 1, p = pS)
pK
#> [1] 0.8998871

### NOTE:
# Example from "The Drunkard's Walk: How Randomness Rules Our Lives" continued...
# Result (1): P(X = k) where k = 0 is calculated as choose(n, k)*(p**k)*((1-p)**(n-k))
# P(X = 0) ~= 0.66 => P(X > 0) ~= 1 - 0.66 = 0.33
```

### Example 2: Simulation

Run a simulation on the problem, and return the set of resulting data

``` r
set.seed(1234)
sim_data <- run_simulation(
  iters = 5000,
  trials = 8,
  sample_space = c(0, 1),
  sample_size = 5,
  run_value = 1,
  run_length = 3
  )

tail(sim_data)
#>      iterations applicable_trials prob_of_zero prob_of_ge_one
#> 4995       4995                 0    0.1017017      0.8982983
#> 4996       4996                 2    0.1016813      0.8983187
#> 4997       4997                 0    0.1018611      0.8981389
#> 4998       4998                 1    0.1018407      0.8981593
#> 4999       4999                 1    0.1018204      0.8981796
#> 5000       5000                 0    0.1020000      0.8980000
```

Plot the probability convergence of the simulation results
<p>
<img src="man/figures/README-fig1.png" class="centering" width="80%">
</p>


NEWS.md
# billmillr 0.1.0

* Added a `NEWS.md` file to track changes to the package.
* Created package scaffolding: functions & documentation.

# billmillr 0.2.0

* Added functions & documentation for the simulation solution.

# billmillr 0.2.1

* Converted analytics functions to use lower snake case.
* Updated documentation based on rstudio::conf(2022) pkg development workshop.

# billmiller 0.2.5

* Built and published pkgdown site to GitHub pages: https://mghoff.github.io/billmillr/
* Function Updates:
  * Added argument checks to functions.
  * Stripped defaults from mandatory function arguments.
  * Optimized for performance.
  * Abstracted simulation set of functions to take either numeric or character args.
  * Added sample_probs arg to simulation set of functions:
    * custom vector of sample probabilities to go alongside the provided sample space.
  * Added `prob_of_ge_one` column to `run_simulation()` output data frame.
* README Updates:
  * Re-wrote narrative/motivation for the package.
  * Stripped out bottom section with advent of `prob_of_ge_one`.


R/simulr.R

#' Find the Longest Run
#'
#' Finds the single longest run ("streak") of the value "1"
#' provided in a binomial sample of size n.
#'
#' @param binom_sample, numeric vector - a binomial sample; e.g. \code{sample(c(0, 1), size = 5, replace = TRUE)}
#'
#' @return number, representing the longest streak (run of "1") in the provided sample
#' @export
#'
#' @examples
#' find_longest_run(binom_sample = sample(c(0, 1), size = 25, replace = TRUE))
find_longest_run <- function(binom_sample) {
  run <- max_run <- 0
  for (i in 2:length(binom_sample)) {
    if (binom_sample[i] == 1 && binom_sample[i] == binom_sample[i-1]) {
      run <- run + 1
      max_run <- max(max_run, run)
    } else {
      run <- 0
      max_run <- max(max_run, run)
    }
  }
  return(max_run + 1)
}


#' Count Runs
#'
#' Out of a given number of trials of flipping a coin n times, count the number of
#' times a run/streak of length k occurs
#'
#' @param trials number, number of attempts to flip a coin n times
#' @param flips number, number of sequential coin flips in which a streak could be counted
#' @param streak_length number, length of streak of interest
#'
#' @return number
#' @export
#'
#' @examples
#' count_runs()
count_runs <- function(trials = 1000, flips = 40, streak_length = 15) {
  x <- NULL
  total_applicable_runs <- 0
  for (t in 1:trials) {
    x[t] <- find_longest_run(binom_sample = sample(c(0, 1), size = flips, replace = TRUE))
    if (x[t] >= streak_length) {
      total_applicable_runs <- total_applicable_runs + 1
    }
  }
  return(total_applicable_runs)
}


#' Run Simulation
#'
#' Iterate over \code{count_runs()}, counting the number of applicable trials where
#' such a trial is defined as having the occurrence of a streak of at least the
#' designated length. Additionally, keep a running tally of the cumulative likelihood
#' of obtaining exactly zero streaks.
#'
#' This function outputs a data frame with three columns:
#' 1.) the iteration number,
#' 2.) the count of applicable trials in that iteration, and
#' 3.) the cumulative probability of obtaining zero streaks.
#'
#' @param iters number, number of sumulation iterations to consider
#' @param trials number, see \code{count_runs()}
#' @param flips number, see \code{count_runs()}
#' @param streak_length number, see \code{count_runs()}
#'
#' @return data frame
#' @export
#'
#' @examples
#' tail(run_simulation())
run_simulation <- function(iters = 100, trials = 1000, flips = 40, streak_length = 15) {
  d <- data.frame(stringsAsFactors = FALSE)
  for (i in 1:iters) {
    d[i, 1] <- i
    d[i, 2] <- count_runs(trials = trials, flips = flips, streak_length = streak_length)
    d[i, 3] <- round(nrow(d[which(d[, 2] == 0), ]) / nrow(d), 6)

  }
  names(d) <- c("iteration", "applicable_trials", "prob_of_zero")
  return(d)
}


R/analytics.R
#' Calculate the Odds of a Streak
#'
#' Calculate the odds of a single streak of at least k heads out of n coin tosses given
#' the probability of heads (tails) is p (q = 1-p).
#'
#' @param num_coins integer, total number of coin flips
#' @param min_heads integer, minimum number of heads to to obtained in a single streak
#' @param prob_heads number, probability of obtaining a heads on a single coin toss
#' @param .saved data frame, contains previously computed probabilities for a given index;
#' used to speed up recursion so not to recalculate every probability should an already seen
#' scenario of num_coins, min_heads, and prob_heads arises.
#'
#' @return numeric, odds of streak
#' @export
#'
#' @examples
#' odds_of_streak(num_coins = 5, min_heads = 3, prob_heads = 0.5)
odds_of_streak <- function(num_coins, min_heads, prob_heads, .saved = NULL) {
  # Computes the probability (i.e. S[n, k]) of getting a run of min_heads
  # i.e. the odds of getting K or more heads in a row out of
  # N independent coin tosses where p is the probability of getting a Heads
  # and q = 1-p is the probability of getting a Tails.

  # Using recursion: S[N, K] = p^K + sum_{j=1, K} ( p^(j-1) * (1-p) * S[N-j, K] )
  # As well as base cases: 1.) S[0, K] = 0 and 2.) S[N, K] = 0
  if (is.null(.saved)) {
    .saved <- data.frame(
      ID = character(),
      result = numeric(),
      stringsAsFactors = FALSE
    )
  }

  # Get a unique identifier for computed value
  ID <- paste0(num_coins, min_heads, prob_heads)

  # If it has been computed before, return pre-computed value
  if (ID %in% .saved$ID) {
    return(as.numeric(.saved[which(.saved$ID == ID), 2]))
    # Handle cases where we have no coins, or more heads than coins
  } else {
    if (min_heads > num_coins | num_coins <= 0) {
      result <- 0
    } else {
      # Use our recursive relationship to compute S[n, k] by breaking it into
      # a sum of terms involving S[n-j, k] for 1 <= j <= k
      result <- prob_heads**min_heads # S[n, k] = p^k + ...
      # S[n, k] = ... + sum_{j=1, k} p^(j-1) * (1-p) * S[n-j, k]
      for (first_tail in seq(1, min_heads + 1)) {
        pr <- odds_of_streak(num_coins - first_tail, min_heads, prob_heads, .saved)
        result <- result + (prob_heads**(first_tail - 1)) * (1 - prob_heads) * pr
      }
      # Save computed value for later use, if needed; i.e. if prob has already been calculated
      .saved <- rbind.data.frame(.saved, as.data.frame(cbind(ID, result)))
    }
  }
  # Return final result
  return(result)
}


#' Odds of at least K streaks
#'
#' Calculate the probability of the occurrence of at least K-streaks out of N-events
#' given the odds of stated streak.
#'
#' @param N integer, total number of coin tossers.
#' @param K integer, number of streaks.
#' @param p number, value between 0 & 1 representing the probability of a single streak.
#'
#' @return a number between 0 and 1
#' @export
#'
#' @examples
#' prob_of_at_least_k(
#'  N = 8,
#'  K = 1,
#'  p = odds_of_streak(num_coins = 5, min_heads = 3, prob_heads = 0.5)
#' )
prob_of_at_least_k <- function(N, K, p) {
  L <- 0
  for (k in K:N) {
    L <- L + choose(N, k) * (p**k) * ((1 - p)**(N - k))
  }
  return(L)
}


R/billmillr.R
#' Odds of a single streak
#'
#' Calculate the odds of a single streak of at least k heads out of n coin tosses given
#' the probability of heads (tails) is p (q = 1-p).
#'
#' @param numCoins integer, total number of coin flips
#' @param minHeads integer, minimum number of heads to to obtained in a single streak
#' @param probHeads number, probability of obtaining a heads on a single coin toss
#' @param .saved data frame, contains previously computed probabilities for a given index;
#' used to speed up recursion so not to recalculate every probability should an already seen
#' scenario of numCoins, minHeads, and probHeads arises.
#'
#' @return numeric, odds of streak
#' @export
#'
#' @examples
#' oddsOfStreak(numCoins = 10, minHeads = 3, probHeads = 0.5)
oddsOfStreak <- function(numCoins = 10, minHeads = 5, probHeads = 0.5, .saved = NULL) {
  # Computes the probability (i.e. S[n, k]) of getting a run of minHeads
  # i.e. the odds of getting K or more heads in a row out of
  # N independent coin tosses where p is the probability of getting a Heads
  # and q = 1-p is the probability of getting a Tails.

  # Using recursion: S[N, K] = p^K + sum_{j=1, K} ( p^(j-1) * (1-p) * S[N-j, K] )
  # As well as base cases: 1.) S[0, K] = 0 and 2.) S[N, K] = 0
  if (is.null(.saved)) {
    .saved <- data.frame(
      ID = character(),
      result = numeric(),
      stringsAsFactors = FALSE
    )
  }

  # Get a unique identifier for computed value
  ID <- paste0(numCoins, minHeads, probHeads)

  # If it has been computed before, return pre-computed value
  if (ID %in% .saved$ID) {
    return(as.numeric(.saved[which(.saved$ID == ID), 2]))
    # Handle cases where we have no coins, or more heads than coins
  } else {
    if (minHeads > numCoins | numCoins <= 0) {
      result <- 0
    } else {
      # Use our recursive relationship to compute S[n, k] by breaking it into
      # a sum of terms involving S[n-j, k] for 1 <= j <= k
      result <- probHeads**minHeads # S[n, k] = p^k + ...
      # S[n, k] = ... + sum_{j=1, k} p^(j-1) * (1-p) * S[n-j, k]
      for (firstTail in seq(1, minHeads + 1)) {
        pr <- oddsOfStreak(numCoins - firstTail, minHeads, probHeads, .saved)
        result <- result + (probHeads**(firstTail - 1)) * (1 - probHeads) * pr
      }
      # Save computed value for later use, if needed; i.e. if prob has already been calculated
      .saved <- data.frame(rbind(.saved, cbind(ID, result)), stringsAsFactors = FALSE)
    }
  }
  # Return computed value
  return(result)
}


#' Odds of at least K streaks
#'
#' Calculate the probability of the occurrence of at least K-streaks out of N-events
#' given the odds of stated streak.
#'
#' @param N integer, total number of coin tossers
#' @param K integer, number of streaks
#' @param P number, probability of a single streak
#'
#' @return a number
#' @export
#'
#' @examples
#' probOfAtLeastK(N = 100, K = 1, P = oddsOfStreak())
probOfAtLeastK <- function(N, K, P) {
  n <- N
  k <- K
  p <- P
  l <- L <- 0
  for (k in K:N) {
    l <- choose(n, k) * (p**k) * ((1 - p)**(n - k))
    L <- L + l
  }
  return(L)
}


R/simulations.R
#' Calculate the Longest Run
#'
#' Finds the single longest run (streak) of a specified value
#' contained within a provided vector
#'
#' @param sample numeric or character vector.
#' @param run_value number or character, the value for which to count sequential occurrences.
#'
#' @return number, the longest run (streak) of values
#' @export
#'
#' @examples
#' s <- sample(c(0, 1), size = 10, replace = TRUE)
#' print(s)
#'
#' find_longest_run(sample = s, run_value = 1)
find_longest_run <- function(sample, run_value) {
  if (!(run_value %in% sample)) {
    return(0)
  } else {
    run <- max_run <- 1
    for (i in 2:length(sample)) {
      if (sample[i] == run_value && sample[i] == sample[i - 1]) {
        run <- run + 1
        max_run <- max(max_run, run)
      } else {
        run <- 1
        max_run <- max(max_run, run)
      }
    }
    return(max_run)
  }
}


#' Count Runs
#'
#' Out of a given number of trials flipping a fair coin (p = 0.5) n times, count the number of
#' times a run (streak) of length k occurs
#'
#' @param trials number, number of times flipping a coin n times.
#' @param sample_space vector, vector of unique values from which to sample.
#' @param sample_probs vector, vector of probabilities equal in length to \code{sample_space}.
#' @param sample_size number, size of sample to generate from \code{sample_space}; e.g. n coin flips.
#' @param run_value number or character, the value on which to count sequential occurrences.
#' @param run_length number, length of streak of interest value.
#'
#' @return number, the total number of applicable runs; i.e. the number of runs where
#' the length of the run is greater than or equal to the specified \code{run_length}.
#' @export
#'
#' @examples
#' count_runs(
#'  trials = 8,
#'  sample_space = c(0, 1),
#'  sample_size = 5,
#'  run_value = 1,
#'  run_length = 3
#' )
count_runs <- function(trials,
                       sample_space,
                       sample_probs = NULL,
                       sample_size,
                       run_value,
                       run_length) {
  if (run_length > sample_size) stop("run_length must be less than or equal to sample_size.")
  if (!(run_value %in% sample_space)) stop("run_value must be a member of the sample_space.")
  if (!is.null(sample_probs)) {
    if (length(sample_probs) != length(sample_space)) stop("sample_probs must be same length as sample_space.")
    if (sum(sample_probs) != 1) stop("sample_probs must sum to 1.")
  }

  total_applicable_runs <- 0
  for (t in 1:trials) {
    lr <- find_longest_run(
      sample = sample(sample_space, size = sample_size, replace = TRUE, prob = sample_probs),
      run_value = run_value
    )
    if (lr >= run_length) {
      total_applicable_runs <- total_applicable_runs + 1
    }
  }
  return(total_applicable_runs)
}


#' Run Simulation
#'
#' Iterate over \code{count_runs()}, counting the number of applicable trials where
#' such a trial is defined as having the occurrence of a streak of at least the
#' designated length. Additionally, keep a running tally of the cumulative likelihood
#' of obtaining exactly zero streaks.
#'
#' This function outputs a data frame with four columns:
#' 1.) the iteration number,
#' 2.) the count of applicable trials in that iteration, and
#' 3.) the cumulative probability of obtaining zero streaks.
#' 4.) the cumulative probability of obtaining at least one streak.
#'
#' @param iters number, number of simulation iterations.
#' @param trials number, see \code{count_runs()}.
#' @param sample_space vector, see \code{count_runs()}.
#' @param sample_probs vector, see \code{count_runs()}.
#' @param sample_size number, see \code{count_runs()}.
#' @param run_value number, see \code{count_runs()}.
#' @param run_length number, see \code{count_runs()}.
#'
#' @return data frame with
#' * 4 columns: \code{iterations}, \code{applicable_trials}, \code{prob_of_zero}, & \code{prob_of_ge_one}; and
#' * the number of rows to be determined by \code{iters}
#' @export
#'
#' @examples
#' tail(
#'  run_simulation(
#'    iters = 100,
#'    trials = 8,
#'    sample_space = c(0, 1),
#'    sample_size = 5,
#'    run_value = 1,
#'    run_length = 3
#'  )
#' )
run_simulation <- function(iters,
                           trials,
                           sample_space,
                           sample_probs = NULL,
                           sample_size,
                           run_value,
                           run_length) {
  d <- data.frame(
    iterations = 1:iters,
    applicable_trials = rep(NA_real_, iters),
    prob_of_zero = rep(NA_real_, iters),
    prob_of_ge_one = rep(NA_real_, iters),
    stringsAsFactors = FALSE
  )
  for (i in 1:iters) {
    d[i, 2] <- count_runs(
      trials = trials,
      sample_space = sample_space,
      sample_probs = sample_probs,
      sample_size = sample_size,
      run_value = run_value,
      run_length = run_length
    )
    d[i, 3] <- nrow(d[which(d[, 2] == 0), ]) / i
    d[i, 4] <- 1 - d[i, 3]
  }
  return(d)
}


bill-miller.R
# Clear Environment
rm(list=ls())

# Define recursive function for calculating odds of streak
probOfStreak <- function(numCoins, minHeads, headProb, saved = NULL) {
    # Computes the probability (i.e. S[n,k]) of getting a run of minHeads
    # i.e. K of more heads in a row out of numCoins
    # i.e. N independent coin tosses where the probability of Heads is p
    # and the probability of Tails is q=1-p
    #
    # Using recursion: S[N,K] = p^K + sum_{j=1,K} ( p^(j-1) * (1-p) * S[N-j,K] )
    # As well as base cases: 1. S[0,K] = 0 and S[N,K] = 0
    # saved <- NULL
    if (is.null(saved)) { saved = data.frame(ID = character(), 
                                            result = numeric(), 
                                            stringsAsFactors = F) 
    }
    
    # Get a unique identifier for computed value
    ID <- paste0(numCoins, minHeads, headProb)
    
    # If it has been computed before, return precomputed value
    if (ID %in% saved$ID) { 
        return(as.numeric(saved[which(saved$ID == ID), 2])) 
    # Handle cases where we have no coins, or more heads than coins
    } else if (minHeads > numCoins | numCoins <= 0) { 
        result <- 0
    # Use our recursive relationship to compute S[n,k] by breaking it into
    # a sum of terms involving S[n-j,k] for 1 <= j <= k
        } else {
            result <- headProb**minHeads # S[n,k] = p^k + ...
            # S[n,k] = ... + sum_{j=1,k} p^(j-1)*(1-p)*S[n-j,k]
            for (firstTail in seq(1, minHeads+1)) {
                pr <- probOfStreak(numCoins-firstTail, minHeads, headProb, saved)
                result <- result + (headProb**(firstTail-1))*(1-headProb)*pr
            }
        # Save computed value for later use if needed
        saved <- data.frame(rbind(saved, cbind(ID, result)), stringsAsFactors = F)
        }
        # Return computed value
        return(result)
    }
### Simple Test
probOfStreak(10, 5, 0.5)

### Example from `The Drunkard's Walk: How Randomness Rules Our Lives`
# tictoc::tic()
# probOfStreak(40, 15, 0.5) # p = 0.000411981
# tictoc::toc() # 24713.89 sec elapsed (6.865 hours)

### Liklihood of at least 1 person out of 1000 getting a streak of 15 heads in 40 flips
# (1) pdf: P(X = k) = choose(n, k)*(p^k)*((1-p)^(n-k))
# (2) cdf: P(X <= k) = sum_{i=0,k} (1) for i <= k
# (3) P(X > k) = 1 - P(X <= k); i.e. 1 - (2)

# this p (below) comes from the full enumeration of a streak of at least 15 Heads from 40 trials
p <- 0.000411981 # for the @Risk simulation, we got 0.0002
n <- 1000
k <- 0
# Result (1): P(X = k)
choose(n, k)*(p**k)*((1-p)**(n-k)) # P(X = 0) ~= 0.66 => P(X > 0) ~= 1 - 0.66 = 0.33

# Create function to calculate likelihood of getting k or more occurences.
probOfAtLeastK <- function(N, K, P) {
    n <- N; k <- K; p <- P
    L = 0; l = 0
    for (k in K:N) {
        l <- choose(n, k)*(p**k)*((1-p)**(n-k))
        L <- L + l
    }
    return(L)
}
probOfAtLeastK(n, 1, p)
